src/grid/cubic.py

```diff
@@ -0,0 +1,516 @@
+import numpy as np
+from scipy.interpolate import CubicSpline
+
+from grid.basegrid import Grid, OneDGrid
+
+
+class CubicGrid(Grid):
+    r"""
+
+    Attributes
+    ----------
+
+    Methods
+    -------
+
+
+    """
+    def __init__(self, oned_grids):
+        if not isinstance(oned_grids, list):
+            raise TypeError("oned_grid should be of type list.")
+        if not np.all([isinstance(grid, OneDGrid) for grid in oned_grids]):
+            raise TypeError("Grid in oned_grids should be of type `OneDGrid`.")
+        if not len(oned_grids) == 3:
+            raise ValueError(
+                "There should be three One-Dimensional grids in `oned_grids`."
+            )
+
+        self._num_pts = tuple([grid.size for grid in oned_grids])
+        self._dim = len(oned_grids)
+
+        # Construct 3D set of points
+        points = np.vstack(
+            np.meshgrid(
+                oned_grids[0].points, oned_grids[1].points, oned_grids[2].points,
+                indexing="ij",
+            )
+        ).reshape(3,-1).T
+        weights = np.kron(
+            np.kron(oned_grids[0].weights, oned_grids[1].weights), oned_grids[2].weights
+        )
+        super().__init__(points, weights)
+
+    @property
+    def num_pts(self):
+        r"""Return number of points in each direction."""
+        return self._num_pts
+
+    @property
+    def dim(self):
+        r"""Return the dimension of the cubic grid."""
+        return self._dim
+
+    @property
+    def origin(self):
+        r"""Return the origin of the coordinate system."""
+        # Bottom, Left-Most, Down-most point of the Cubic Grid in [-1, 1]^3.
+        return self.points[0]
+
+    def integrate(self, *value_arrays, is_zero_boundary=False, type="Rectangle"):
+        r"""
+        Integrate any real-valued function on Euclidean space.
+
+        Parameters
+        ----------
+        *value_arrays : (np.ndarray(N, dtype=float),)
+            One or multiple value array to integrate, these are usally the functions evaluated
+            at the points of the grid.
+        is_zero_boundary : bool, optional
+            If the function being integrated is zero on the boundary, then this changes the
+            weight used.
+        type : str, optionla
+            The type of weights used, either "Rectangle", or "Fourier".
+
+        Returns
+        -------
+        float :
+            Return the integration of the function.
+
+        Raises
+        ------
+        TypeError
+            Input integrand is not of type np.ndarray.
+        ValueError
+            Input integrand array is given or not of proper shape.
+
+        """
+        return super().integrate(*value_arrays)
+
+    def interpolate_function(
+            self, real_pt, func_values, use_log=False, nu_x=0, nu_y=0, nu_z=0,
+    ):
+        r"""
+        Interpolate function at a point.
```

Comment:
@tczorro is refactoring the `interpolation` module, so I am wondering whether this is sth that should be added there.   Please check with him @Ali-Tehrani.